### PR TITLE
Fix included destination tax

### PR DESCRIPTION
**Pull request by SteveIreland**
This pull request adds missing support for ""tax included" when configured for destination based taxing.

### CODE REVIEW

**Comment by odiea** [comment]
I tried this fix but it appears to have an issue.
phpmyadmin db
2020-04-05 06_13_20-Window
Then this is what I see after loading the Taxes Module. I have tried deleting them but they just return again.
2020-04-05 06_13_45-Window

**Comment by SteveIreland** [comment]
Thanks @odiea I'll try to see what it takes to replicate the issue.

File: application/libraries/Tax_lib.php [hunk 1]
```diff
@@ -150,9 +150,10 @@ public function get_taxes(&$cart)
 
 	public function get_included_tax($quantity, $price, $discount_percentage, $discount_type, $tax_percentage, $tax_decimal, $rounding_code)
 	{
-		$tax_amount = $this->CI->sale_lib->get_item_tax($quantity, $price, $discount_percentage, $discount_type, $tax_percentage);
-
-		return Rounding_mode::round_number($rounding_code, $tax_amount, $tax_decimal);
+		$item_total = $this->CI->sale_lib->get_item_total($quantity, $price, $discount_percentage, $discount_type, TRUE);
```
Comment (jekkos) [inline application/libraries/Tax_lib.php:153]: @SteveIreland would be cool if you could add a small unit test for this one ! check example i did for the barcode_lib

Comment (SteveIreland) [reply application/libraries/Tax_lib.php:153]: Okay, I'll give it a shot.  I like examples!

**Comment by SteveIreland** [comment]
@odiea It's been a little busy this weekend for me but I've been giving some thought to the problem you reported and I need more information.  Your issue looks like a data migration issue. Did you migrate your test data from a non-destination based tax to a destination based tax system? What was the sequence of steps to make that happen because it certainly looks like the migration is the failure point in your example.

**Comment by odiea** [comment]
Ok Here goes.
When doing the update to multi currency the first part of the script works but not the second part.

INSERT INTO `ospos_app_config` (`key`, `value`) VALUES
('last_exchange_rate', '1.00'),
('number_locale_alt', 'en_US'),
('use_alternate_currency', '0'),
('currency_symbol_alt', '$');

I had to manually add this part. 
ALTER TABLE `ospos_sales`
	ADD COLUMN `exchange_rate` decimal(28,14) DEFAULT 1.0 AFTER `sale_type`,
	ADD COLUMN `number_locale_alt` varchar(10) DEFAULT '' AFTER `exchange_rate`,
	ADD COLUMN `currency_symbol_alt` varchar(3) DEFAULT '' AFTER `number_locale_alt`;

Then when creating the 1st sale the following errors happen. Don't know if you should put in a default value?

I did the upgrade on a different db and the issue with multiple tax codes does not show.

ERROR - 2020-04-13 04:06:59 --> Severity: Warning --> A non-numeric value encountered C:\wamp64\www\multi3\application\views\sales\register.php 165
ERROR - 2020-04-13 04:06:59 --> Severity: Warning --> Division by zero C:\wamp64\www\multi3\application\views\sales\register.php 165
ERROR - 2020-04-13 04:06:59 --> Severity: Warning --> A non-numeric value encountered C:\wamp64\www\multi3\application\views\sales\register.php 215
ERROR - 2020-04-13 04:06:59 --> Severity: Warning --> Division by zero C:\wamp64\www\multi3\application\views\sales\register.php 215
ERROR - 2020-04-13 04:07:50 --> Query error: Column 'value' cannot be null - Invalid query: INSERT INTO `ospos_app_config` (`key`, `value`) VALUES ('last_exchange_rate', NULL)
ERROR - 2020-04-13 04:11:50 --> Query error: Column 'value' cannot be null - Invalid query: INSERT INTO `ospos_app_config` (`key`, `value`) VALUES ('last_exchange_rate', NULL)
ERROR - 2020-04-13 04:12:50 --> Severity: Warning --> A non-numeric value encountered C:\wamp64\www\multi3\application\views\sales\register.php 165
ERROR - 2020-04-13 04:12:50 --> Severity: Warning --> Division by zero C:\wamp64\www\multi3\application\views\sales\register.php 165
ERROR - 2020-04-13 04:12:50 --> Severity: Warning --> A non-numeric value encountered C:\wamp64\www\multi3\application\views\sales\register.php 215
ERROR - 2020-04-13 04:12:50 --> Severity: Warning --> Division by zero C:\wamp64\www\multi3\application\views\sales\register.php 215

I know you have not updated it recently but if this is put in place there are issues with the current locale view and this locale view.

**Comment by jekkos** [comment]
Just pinging the thread here, it would be nice to include this fix in our next minor release.

File: application/libraries/Sale_lib.php [hunk 2]
```diff
@@ -472,29 +472,27 @@ public function get_totals($taxes)
...
 		$totals['prediscount_subtotal'] = $prediscount_subtotal;
 		$totals['total_discount'] = $total_discount;
-		$totals['subtotal'] = $subtotal;
 		$sales_tax = 0;
 
-		foreach($taxes as $tax_excluded)
+		foreach($taxes as $tax)
 		{
-			if($tax_excluded['tax_type'] == Tax_lib::TAX_TYPE_EXCLUDED)
+			if($tax['tax_type'] == Tax_lib::TAX_TYPE_EXCLUDED)
```
Comment (objecttothis) [inline application/libraries/Sale_lib.php:484]: I think you probably should use === here rather than ==.

Comment (objecttothis) [reply application/libraries/Sale_lib.php:484]: https://stackoverflow.com/questions/3084580/the-importance-of-using-instead-of-in-php

**Comment by objecttothis** [comment]
@SteveIreland I did a code review.  I also added 3.3.2 as the release target... Do you think it should get punted to 3.4.0?  If it needs to go in 3.3.2, how soon do you think you can get this done so that we can release 3.3.2?

File: application/libraries/Sale_lib.php (hunk 2 above)
Comment (SteveIreland) [reply application/libraries/Sale_lib.php:484]: I made the requested change.

Comment (objecttothis) [reply application/libraries/Sale_lib.php:484]: > 
> 
> I made the requested change.

Did you push the commit to the branch?  I just see the initial commit in April.

Comment (SteveIreland) [reply application/libraries/Sale_lib.php]: I made the change but discovered during testing that the change broke it ... but not wanting to revert that change I decided to fix the reason it was broken.

**Comment by SteveIreland** [comment]
@objecttothis has this merge block and I'm unable to figure a way around the block.  So it remains unmerged.

This fix also includes a temporary hack that corrects a bad reference to an unrelated JavaScript file which might possibly resolve the issue reported on #2850 

And finally I will start diving into the issue that @odiea reported to see if I can understand and replicate it.

**Comment by odiea** [comment]
@SteveIreland . I deleted the entries from the db and the issue was resolved so don't worry unless someone else has an issue.

**Comment by objecttothis** [comment]
> 
> 
> @objecttothis has this merge blocked and I'm unable to figure a way around the block. So it remains unmerged.
> 
> This fix also includes a temporary hack that corrects a bad reference to an unrelated JavaScript file which might possibly resolve the issue reported on #2850
> 
> And finally I will start diving into the issue that @odiea reported to see if I can understand and replicate it.

Sorry, I'm GMT+4 so I didn't approve and merge it until just now when I started work.  To clarify, the system just requires one code review and it doesn't have to be the person that requested changes initially.  AFAIK I don't have any special abilities to lock the PR from being merged if another dev approves it.

**Pull request merged**